fairseq/sequence_generator.py

```diff
@@ -367,6 +369,10 @@ def calculate_banned_tokens(bbsz_idx):
                     # handle prefixes of different lengths
                     partial_prefix_mask = prefix_tokens[:, step].eq(self.pad)
                     if partial_prefix_mask.any():
```

Comment:
Thanks for the fix!
But seems like it can only cover cases where prefixes of different lengths are present.

Here are repro steps to test a case of single sentence with prefix.
1) Download pre-trained wikitext LM here https://github.com/pytorch/fairseq/blob/master/examples/language_model/README.md (I stored it to ~/downloads/newwiki locally)
2) run
```
python fairseq_cli/interactive.py ~/downloads/newwiki/private/home/abaevski/data/wiki103/ --path ~/downloads/newwiki/wiki103.pt --task='language_modeling' --nbest 5
...
| Type the input sentence and press return:
hello world
H-0     -5.840474605560303      hello world ! "
P-0     -17.4183 -8.5166 -1.1164 -1.6743 -0.4767
H-0     -5.840474605560303      hello world ! "
P-0     -17.4183 -8.5166 -1.1164 -1.6743 -0.4767
H-0     -5.840474605560303      hello world ! "
P-0     -17.4183 -8.5166 -1.1164 -1.6743 -0.4767
H-0     -5.840474605560303      hello world ! "
P-0     -17.4183 -8.5166 -1.1164 -1.6743 -0.4767
H-0     -5.840474605560303      hello world ! "
P-0     -17.4183 -8.5166 -1.1164 -1.6743 -0.4767
```

I run these steps with this fix, but generated candidates are the same. We have to adjust the fix to cover this case as well.

Reply:
Hi @okhonko , thx for your review. That's a bug indeed, because the previous fix doesn't deal with longest prefix_tokens (when step==prefix_tokens.size(1)). Therefore, I fix it by running another step (when step==prefix_tokens.size(1)), and force partial_prefix_mask to all ones.

btw, I found out that those codes dealing with prefix_tokens don't work with other search methods (e.g. sampling), which is due to the cand_size / beam_size mismatch. I believe additional fix would change other codes, so I leave it here with an "assertion".